storagestate/storagestate.go

```diff
@@ -0,0 +1,783 @@
+	storageDir          string
+	stateDir            string
+	storage             Storage
+	statesMap           map[cloudprotocol.InstanceIdent]*stateParams
+	watcher             *fsnotify.Watcher
+	stateChangedChannel chan StateChangedInfo
+	newStateChannel     chan cloudprotocol.NewState
+	stateRequestChannel chan cloudprotocol.StateRequest
+	isSamePartition     bool
+}
+
+type stateParams struct {
+	instanceID         string
+	stateDir           string
+	stateFilePath      string
+	quota              uint64
+	checksum           []byte
+	changeTimer        *time.Timer
+	changeTimerChannel chan bool
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+// New creates storagestate instance.
+func New(storageDir string, stateDir string, storage Storage) (storageState *StorageState, err error) {
+	log.Debug("Create storagestate")
+
+	storageState = &StorageState{
+		storage:             storage,
+		storageDir:          storageDir,
+		stateDir:            stateDir,
+		statesMap:           make(map[cloudprotocol.InstanceIdent]*stateParams),
+		stateChangedChannel: make(chan StateChangedInfo, stateChannelSize),
+		newStateChannel:     make(chan cloudprotocol.NewState, stateChannelSize),
+		stateRequestChannel: make(chan cloudprotocol.StateRequest, stateChannelSize),
+	}
+
+	if err = checkAndCreateDir(storageState.storageDir); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if err = checkAndCreateDir(storageState.stateDir); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if storageState.watcher, err = fsnotify.NewWatcher(); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	storageMountPoint, err := fs.GetMountPoint(storageDir)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	stateMountPoint, err := fs.GetMountPoint(stateDir)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	storageState.isSamePartition = storageMountPoint == stateMountPoint
+
+	go storageState.processWatcher()
+
+	return storageState, nil
+}
+
+// Close close storagestate instance.
+func (storageState *StorageState) Close() {
+	log.Debug("Close storagestate")
+
+	storageState.watcher.Close()
+}
+
+// Setup setup storagestate instance.
+func (storageState *StorageState) Setup(
+	instanceID string, params SetupParams,
+) (storagePath string, statePath string, stateChecksum []byte, err error) {
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	log.WithFields(log.Fields{
+		"instaceID":    instanceID,
+		"storageQuota": params.StorageQuota,
+		"stateQuota":   params.StateQuota,
+	}).Debug("Setup storage and state")
+
+	currentStorageState, err := storageState.storage.GetStorageStateInfoByID(instanceID)
+	if err != nil {
+		return "", "", nil, aoserrors.Wrap(err)
+	}
+
+	var newStorageState StorageStateInstanceInfo
+
+	defer func() {
```

Comment:
If startStateWatching is the last op in this function, no need for this defer.

Reply:
Done